### DISCUSSION TITLE
Assertion reuse

### DISCUSSION THREAD

**Discussion by shfrmn** [tstyche/tstyche#167] [Ideas]
Hi! Thank you for this fantastic library, it's an absolute joy to use. Can't believe you reached this level of DX and documentation quality with so few Github stars. Hope that's about to change!

---

### Intro

I would like to bring up the topic of assertion reuse.  

The need to group and reuse assertions arises naturally. However at first I struggled to elaborate it in a way that would really explain the essence of the issue.

This is because for the most part this need is covered by TypeScript interfaces and types as such. For example, if I wanted to check the type of object properties A and B, I could make a single assertion `.toEqual<{a: A, b: B}>()` instead of making two individual assertions. So at first it might seem like grouping and reusing assertions is not at all necessary.

While it's perfectly reasonable for users to write 10 property assertions instead of just one interface assertion, I wanted to come up with a more complex example that would illustrate an edge case when covering everything by one assertion is very hard. The main point of the example is mostly to show the unnecessary cognitive load resulting from trying to express a complex assertion in one type, when it could just be a lot of individual assertions of much simpler types.

<details>
<summary><b>Counter-example</b></summary>

```typescript
type Wrapper<T> = {value: T}

type Recursive<T> = T extends () => any
  ? () => Wrapper<Recursive<T>>
  : T extends Record<any, any>
    ? {[K in keyof T]: Wrapper<Recursive<T[K]>>}
    : T
```

What makes this snippet relevant is that if our goal was to write tests for `Recursive`, without ever referring to the type itself, it could only be done using `.toMatch()` and `.toBeAssignable()`, but never `.toEqual()`. The reason for this is that the structure of `Recursive<T>` is too complex for TypeScript to fully reduce to its resolved form, even given a specific `T`.

```typescript
interface Greeting {
  hello: {
    world: "hi"
  }
}

interface ResolvedGreeting {
  hello: {
    value: {
      world: {
        value: string
      }
    }
  }
}

// will not pass
expect<Recursive<Greeting>>().type.toEqual<ResolvedGreeting>()
expect<Recursive<Greeting>>().type.toBeAssignable<ResolvedGreeting>()

// will pass
expect<Recursive<Greeting>>().type.toMatch<ResolvedGreeting>()
expect<ResolvedGreeting>().type.toBeAssignable<Recursive<Greeting>>()

```

Now, if we wanted to assert that some random property `covfefe` doesn't exist in `Recursive<Greeting>` there are only three ways to do that:
- `expect<Recursive<Greeting>>().not.toHaveProperty("covfefe")`
- `expect<Recursive<Greeting>>().type.not.toMatch<ResolvedGreeting & {covfefe: any}>()`
- `expect<ResolvedGreeting & {covfefe: any}>().type.not.toBeAssignable<Recursive<Greeting>>()`

All of these can only be expressed separately from the main interface assertion.

</details>

### Solutions

I see this as a kind of fork in the road.

One path from here is to allow users to group and reuse assertions any way they want. The benefit of going this way is that it greatly reduces cognitive load while writing tests and reduces the need to copypaste.

Another path is to expose additional assertion methods to address specific frictions, such as the one I showcased in the example above. This may be easier to implement, but will not fundamentally resolve the issue of having to copypaste groups of assertions, because I believe generally users will follow the path of least resistance and write a lot of individual assertions anyway.

### Possible solutions

<details>
<summary><b>Assertion reuse</b></summary>

In regular functional tests, assertion reuse can be achieved using functions.

```typescript
// dummy test
function assertRecursive(value: Recursive<Greeting>) {
  expect(value).type.toMatch<ResolvedGreeting>()
  expect(value).not.toHaveProperty("covfefe")
}

test("Use case 1", () => {
  assertRecursive(valueA)
})

test("Use case 2", () => {
  assertRecursive(valueB)
})
```

But of course this will never work at the type level, since this will just typecheck `assertRecursive` function itself (once).

One way to solve this would be by introducing a higher level combinator (I suspect type inference will be tricky here):

```typescript
const assertInvariant = invariant((value) => [
  expect(value).type.toMatch<ResolvedGreeting>(),
  expect(value).not.toHaveProperty("covfefe")
])

test("Use case", () => {
  assertInvariant(value)
})
```
</details>

<details>
<summary><b>Addressing my example specifically</b></summary>

Perhaps a new `.toMatchStrict()` method could be exposed that would prohibit additional properties. 

Another way is to provide documentation for something like:
```typescript
expect<Exclude<Recursive<Greeting>>, {covfefe: any}>().type.toMatch<ResolvedGreeting>()
```

Or a bit simpler:

```typescript
type WithoutProps<T, Props> = Exclude<T, {[P in Props]: any}>

expect<WithoutProps<Recursive<Greeting>>, 'covfefe'>().type.toMatch<ResolvedGreeting>()
```

</details>

---

In hindsight, the problem space seems quite simple to me. But then I just spent two hours thinking about it, so I thought I'd post my thoughts here.

#### Takeaways
- I think ideally users of the library should have simple API for combining individual assertions. The cognitive load of creating complex types (e.g. conditional, recursive, covariants and so on) is enough in the first place, there's no need to double the complexity. Testing should stay explicit and easy.
- While `.toEqual`, `.toMatch` and `.toBeAssignable` cover *almost* everything, it's not everything. I propose to add the strict versions of these methods or just a second optional type parameter to perform exclusion.
- Documentation could still be clearer that flipping the left and right sides of `.toMatch` and `.toBeAssignable` makes them interchangeable.
- Even advanced users occasionally get confused with regards to which of their own types are wider/narrower and then consequently whether they should use `.toMatch` or `.toBeAssignable`. Possible ideas:
  
  1. Slightly change descriptions of the methods (indicating which side is supposed to be narrower and which one wider)
  2. Slightly change failed assertion messages to always make it clear which type is wider. "not a subtype" and "not assignable" is good phrasing, but can still use clarification.
  3. If possible, actually show the underlying TypeScript error message (e.g. "property 'field' is missing from type Y", instead of just "X is not assignable to Y" followed by the entire type).

**Comment by mrazauskas** [comment]
Thanks for feedback. Glad you like TSTyche.

I will come back with more detailed response later. Here are quick reaction on the Takeaways:

- current set of assertions do not cover everything what is needed. I am open for proposals, so thanks for talking about your use case.
- as an example, earlier it was pointed out that `.toBeAssignable()` is hard to understand. Hence `.toBeAssignableTo()` and `.toBeAssignableWith()` matchers are replacing it in TSTyche 2 (beta should come out very soon) (see #141).
- hopefully this makes flipping sides of assignability matcher more clear (any feedback on naming is still welcome).
- in the future I also plan to rework the error messages and to make them explain relation just like TypeScript does. That’s rather involving, but doable.

In the other news: currently I am busy with implementing the watch mode. Basic implementation is already in place, I just want to polish a couple of edges. It will ship with upcoming beta.

**Comment by mrazauskas** [comment]
One more detail, I think `.toMatch()` must be reworked. It should take only object and check if subset of properties is identical. 

And the subtype check should be moved to a dedicated matcher. Perhaps even two matchers: `.toBeSubtypeOf()`, `.toBeSupertypeOf()`.

**Comment by mrazauskas** [comment]
I just opened #168 to track behaviour issue of `.toMatch()` separately. The changes I described there would always use excess property checks.

Also I just took another look at documentation of the [Excess Property Checks](https://tstyche.org/reference/expect-api#excess-property-checks). Hm.. It did not sound strange when I was writing those lines, but all that sounds rather unpredictable. What if assignability matchers would always fail, if excess properties are found? (Not sure how to implement that, but I will figure out.)

**Comment by nikelborm** [comment]
I guess my proposal from [here (#465)](https://github.com/tstyche/tstyche/issues/465) is about the same thing raised in this discussion

My idea was to use generics as a way to redirect type information into the `tstyche` method calls like this:

```ts
const expectEquality = <A, B>() => expect<A>().type.toBe<B>();

// should pass
expectEquality<[1, 2 | 3], [1, 2 | 3]>();

// should fail
expectEquality<[1, 2 | 3], [1, 2] | [1, 3]>();
```

**Comment by mrazauskas** [comment]
Indeed, template test files is great way to reuse assertions. (@nikelborm, thanks again for finding the solution.)

For documentation, see: https://tstyche.org/guides/template-test-files
